File: Programmers/[12984] 지형 편집/KingDonggyu/12984.js
function solution(land, P, Q) {
  land = land.flat();

  const getCost = (height) => {
    return land.reduce((acc, cur) => {
      [0, 1].forEach((h, i) => {
        acc[i] += cur < height + h
          ? ((height + h) - cur) * P
          : (cur - (height + h)) * Q;
      })
      return acc;
    }, [0, 0]);
  };

  let answer;
  let left = Math.min(...land);
  let right = Math.max(...land);

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    const [cost1, cost2] = getCost(mid);

    if (cost1 < cost2) {
      right = mid - 1;
      answer = cost1;
      continue;
    }

    left = mid + 1;
    answer = cost2;

    if (cost1 === cost2) {
      break;
    }
  }

  return answer;
}

File: Programmers/[12984] 지형 편집/Johoseong/12984.py
def calc_actions(height, cur, P, Q):
    if cur == height: # 아무거도 안해도 됨
        return 0
    elif cur < height: # 쌓아야 함
        return (height - cur) * P
    else: # 제거해야 함
        return (cur - height) * Q

def calc_cost(height, P, Q, N, land):
    cost_mid, cost_right = 0, 0
    for r in range(N):
        for c in range(N):
            cost_mid += calc_actions(height, land[r][c], P, Q)
            cost_right += calc_actions(height + 1, land[r][c], P, Q)
    return cost_mid, cost_right

def solution(land, P, Q):
    answer = 10 ** 10
    N = len(land)
    left, right = 0, 0
    for l in land:
        right = max(l)

    while left <= right:
        mid = (left + right) // 2
        cost_mid, cost_right = calc_cost(mid, P, Q, N, land)

        if cost_mid == cost_right:
            answer = min(answer, cost_mid)
            break
        elif cost_mid > cost_right:
            answer = min(answer, cost_right)
            left = mid + 1
        else:
            answer = min(answer, cost_mid)
            right = mid - 1

    return answer

print(solution([[1, 2], [2, 3]], 3, 2))